Start to support .loc[slice] (unsure whether to upstream)

jkew opened the pull request:
This helps provide limited support for using a slice in .loc calls; 
which is a pre-requisite to a lot of other interesting operations. 
This only provides support for positional indexes and it does so
by translating to iloc. The next step will be to start providing
slice support for labels properly and then work on the time index.

jkew committed:
Start to support .loc[slice]
This helps provide limited support for using a slice in .loc calls;
which is a pre-requisite to a lot of other interesting operations.
This only provides support for positional indexes and it does so
by translating to iloc. The next step will be to start providing
slice support for labels properly and then work on the time index.

pyrito commented on modin/pandas/indexing.py:
```diff
@@ -633,6 +633,25 @@ def __getitem__(self, key):
+        # support proper labels (sorted) we will need to translate the
+        # existing index to a mask OR translate the expression to a set of
+        # predicates.
+        if is_slice(key):
+            if key.step is not None:
+                raise NotImplementedError("Slice with step not supported for loc")
+            if key.start is None and key.stop is None:
+                return self.df 
+            if key.start is None:
+                new_index = (self.df.index <= key.stop)
+                return self.df.iloc[ new_index.start : ( new_index.stop + 1) ]
+            if key.stop is None:
+                new_index = (self.df.index >= key.start)
+                return self.df.iloc[ new_index.start : ( new_index.stop + 1) ]
+            # Works for a positional mapping
+            return self.df.iloc[ key.start : ( key.stop + 1) ]
```
Is this necessary? IIRC, we should be handling slices without needing this change, right? Or is that support more general than we can support right now?

jkew replied:
No, we currently cannot handle slices at all because we don't have the index support for it. Some of this will change a bit as I implement support for this on non-default indexes.

pyrito replied:
Ahhh I see, so that's why we're just forcing it to do the `iloc` here instead.

pyrito commented on modin/pandas/indexing.py:
```diff
@@ -633,6 +633,25 @@ def __getitem__(self, key):
+            if key.start is None and key.stop is None:
+                return self.df 
+            if key.start is None:
+                new_index = (self.df.index <= key.stop)
```
This is interesting - does this logic match pandas behavior?

jkew replied:
Not exactly, in pandas this operation will produce a binary mask array, the same length as the array. We cannot quite do that here because the array would be the same size as the dataset.

jkew replied:
Once way to emulate this behavior is to provide a compressed "mask" index. Rather than materializing the actual booleans we would return a pandas.arrays.SparseArray

jkew committed:
A more pandas-like interface to ge/le on an index. Not necessarily efficient
because we are not using sparse arrays, but maybe we can figure that out
later.

jkew committed:
Merge branch 'ponder-on-modin-0-19-0' of github.com:ponder-org/modin-public into jkew/to-loc-or-not-to-loc
